lib/std/heap/general_purpose_allocator.zig

```diff
@@ -174,7 +179,7 @@ pub fn GeneralPurposeAllocator(comptime config: Config) type {
         const total_requested_bytes_init = if (config.enable_memory_limit) @as(usize, 0) else {};
         const requested_memory_limit_init = if (config.enable_memory_limit) @as(usize, math.maxInt(usize)) else {};
 
-        const mutex_init = if (config.thread_safe) std.Mutex{} else std.mutex.Dummy{};
+        const mutex_init = if (config.mutex_init) |init| init else if (config.thread_safe) std.Mutex{} else std.mutex.Dummy{};
```

Comment:
Does this even compile? Don't you need `init{}`?

Reply:
it compiles passed the existing tests on zig stdlib.  Did I mess something up here?  If config.mutex_init is defined, then the if will trap init = config.mutex_init.

Reply:
> it compiles and passed the existing tests on zig stdlib.

That doesn't tell much as `mutex_init` is never tested.

>  Did I mess something up here? 

I guess you're going to pass the `erl_drv_mutex_create` value here, no?
This kind of API is not flexible enough, what if the mutex type doesn't like being initialized from some static data?
Right now GPA makes this assumption as the stdlib mutexes satisfy this property, allowing arbitrary mutex types may actually break this invariant.

Moreover there's a hidden footgun here, consider this piece of code:
```zig
var mutex_ptr = erl_drv_mutex_create("foo");
var gpa = GeneralPurposeAllocator(.{.mutex_init = mutex_ptr });
```

Since `mutex_ptr` is a pointer to the mutex the caller has to manually keep track of the two object lifetimes and make sure the mutex object lives at least as much as the gpa object does.

Reply:
this is not a problem for the erlang system.  You would create one mutex per module, erl_drv_mutex_create would live in the module creation hook onload:  http://erlang.org/doc/man/erlang.html#load_nif-2

Reply:
> And that responsibility is on the person who implements their mutex interface

There's no mutex interface, the stdlib ones (and pthread's) just happen to be initializable from static data.

Reply:
agreed there is no "official" mutex interface, but it general interfaces are duck-typed and checked at comptime in zig, so as long as you provide a type that supplies the needed fields it should be fine.

thanks for pushing me towards a correct solution.  I've fixed the logic and added a test showing that this works.